### PR TITLE
feat: withdraw all stake on liquidation

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

### CODE REVIEW

File: parachain/runtime/tests/test_redeem.rs [hunk 1]
```diff
@@ -642,14 +642,13 @@ fn integration_test_redeem_wrapped_cancel_liquidated_reimburse() {
...
             ParachainState::get(),
-            post_liquidation_state.with_changes(|user, vault, liquidation_vault, fee_pool, _| {
+            post_liquidation_state.with_changes(|user, vault, liquidation_vault, _, _| {
                 // to-be-redeemed decreased, forwarding to liquidation vault
                 vault.to_be_redeemed -= redeem.amount_btc + redeem.transfer_fee_btc;
                 liquidation_vault.to_be_redeemed -= redeem.amount_btc + redeem.transfer_fee_btc;
 
                 // tokens are given to the vault, minus a fee that is given to the fee pool
                 vault.free_tokens += redeem.amount_btc + redeem.transfer_fee_btc;
-                fee_pool.vault_wrapped_rewards += vault_rewards(redeem.fee);
```
Comment (sander2) [inline parachain/runtime/tests/test_redeem.rs:652]: Why isn't the fee_pool given a share anymore? What happens to this fee? (also not that the comment above now is incorrect)

File: parachain/runtime/tests/test_redeem.rs [hunk 2]
```diff
@@ -698,10 +697,7 @@ fn integration_test_redeem_wrapped_execute_liquidated() {
         // NOTE: changes are relative the the post liquidation state
         assert_eq!(
             ParachainState::get(),
-            post_liquidation_state.with_changes(|user, vault, liquidation_vault, fee_pool, _| {
-                // fee given to fee pool
-                fee_pool.vault_wrapped_rewards += vault_rewards(redeem.fee);
```
Comment (sander2) [inline parachain/runtime/tests/test_redeem.rs:703]: Like above, why this change?

File: parachain/runtime/tests/test_redeem.rs (hunk 1 above)
Comment (gregdhill) [reply parachain/runtime/tests/test_redeem.rs:652]: It is but in this case when the SLA is taken to zero, all rewards are withdrawn since there is only one vault in this test.

Comment (sander2) [reply parachain/runtime/tests/test_redeem.rs:652]: But only the stake is withdrawn, right? I mean, any earned rewards are not withdrawn? I think withdrawing stake should not change the total amount of money in the fee pool.

Comment (gregdhill) [reply parachain/runtime/tests/test_redeem.rs:652]: Ah yes, so actually once the vault is liquidated it is now removed from the reward pool. Once `cancel_redeem` is called all rewards are distributed to the maintenance fund since `total_stake` is now zero.

Comment (sander2) [reply parachain/runtime/tests/test_redeem.rs:652]: Can you point me to where everything is distributed to the maintainance fund? Also, to be clear, liquidation does not mean the vault loses all rewards it has earned so far, right?

Comment (gregdhill) [reply parachain/runtime/tests/test_redeem.rs:652]: https://github.com/interlay/btc-parachain/blob/master/crates/fee/src/lib.rs#L298

Comment (gregdhill) [reply parachain/runtime/tests/test_redeem.rs:652]: > Also, to be clear, liquidation does not mean the vault loses all rewards it has earned so far, right?

Correct

Comment (sander2) [reply parachain/runtime/tests/test_redeem.rs:652]: > https://github.com/interlay/btc-parachain/blob/master/crates/fee/src/lib.rs#L298

Could we change the distribute function like this?
```
fn distribute(reward: Self::SignedFixedPoint) -> Result<Self::SignedFixedPoint, DispatchError> {
        let total_stake = Self::total_stake();
        // if total_stake.is_zero() {
        //     return Ok(Self::SignedFixedPoint::zero());
        // }

        let reward_div_total_stake = reward
            .checked_div(&total_stake)
            .or(0);
        checked_add_mut!(RewardPerToken<T, I>, &reward_div_total_stake);
        checked_add_mut!(TotalRewards<T, I>, &reward);
        Ok(reward)
    }
    ```
    
I feel like this results in fewer edge cases that we have to test. Right now, you are removing an important test. Either we need to add another test to explicitly check what happens if there _is_ a vault remaining in the system, or we make the change I proposed

Comment (gregdhill) [reply parachain/runtime/tests/test_redeem.rs:652]: @sander2 since investigating distribution without stake is not the purpose of this PR shall we merge and ticket? Or shall I refactor the tests to allow for an additional participant who would receive all rewards upon liquidation?

**Review by sander2 — APPROVED** [review]
